src/main/java/org/sudoku/Sudoku.java

```diff
@@ -1,143 +1,150 @@
-      }
-
-      if(result) continue;
-
-      for (Lane lane : lanes) {
-        Set<Cell> cells = lane.checkCellsWithMatchingNumbers();
-        reorderCells(cells);
-        result = result || !cells.isEmpty();
-      }
-
-      if(result) continue;
-
-      Set<Cell> changedCells = new HashSet<>();
-      for(Entry<Lane, Set<Lane>> squareMap : squaresToLanes.entrySet()){
-        Lane square = squareMap.getKey();
-        for(Lane lane: squareMap.getValue()) {
-          Set<Cell> intersectionCells = square.intersect(lane);
-          Set<Cell> substractionLaneFromSquareCells = square.substract(lane);
-          Set<Cell> substractionSquareFromLaneCells = lane.substract(square);
-
-          Set<Byte> intersectionNumbers = new HashSet<>();
-          Set<Byte> substractionLaneFromSquareNumbers = new HashSet<>();
-          Set<Byte> substractionSquareFromLaneNumbers = new HashSet<>();
-
-          for(Cell cell : intersectionCells) {
-            Set<Byte> possibleNumbers = cell.getPossibleNumbers();
-            intersectionNumbers.addAll(possibleNumbers);
-          }
-
-          for(Cell cell : substractionLaneFromSquareCells) {
-            Set<Byte> possibleNumbers = cell.getPossibleNumbers();
-            substractionLaneFromSquareNumbers.addAll(possibleNumbers);
-          }
-
-          for(Cell cell : substractionSquareFromLaneCells) {
-            Set<Byte> possibleNumbers = cell.getPossibleNumbers();
-            substractionSquareFromLaneNumbers.addAll(possibleNumbers);
-          }
-
-          HashSet<Byte> clonedIntersectionNumbers = new HashSet<>(intersectionNumbers);
-          clonedIntersectionNumbers.removeAll(substractionLaneFromSquareNumbers);
-          if(!clonedIntersectionNumbers.isEmpty()){
-            removeNumbersFromCells(changedCells, substractionSquareFromLaneCells, clonedIntersectionNumbers);
-          }
-          intersectionNumbers.removeAll(substractionSquareFromLaneNumbers);
-          if(!intersectionNumbers.isEmpty()) {
-            removeNumbersFromCells(changedCells, substractionLaneFromSquareCells, intersectionNumbers);
-          }
+
+    public void recalcCell(Cell a){
+        a.possibleValues.clear();
+        for(int i = 1; i < 10; i++){
+            if(fitTest(a, i)){
+                a.possibleValues.add(i);
+            }
         }
-      }
-      reorderCells(changedCells);
-      result = !changedCells.isEmpty();
     }
-  }
 
-  private void checkCellsWithSinglePossibleNumber() {
-    Iterator<Cell> cellIterator = cells.stream().filter(cell -> cell.possibleNumbersSize() == 1).iterator();
+    public boolean fitTest(Cell a, int n){
+        List<Cell> horiz = cells.stream().filter(c->c.getY()==a.getY()).collect(Collectors.toList());
+        List<Cell> vert = cells.stream().filter(c->c.getX()==a.getX()).collect(Collectors.toList());
+        List<Cell> sq = cells.stream().filter(c->c.getSquare()==a.getSquare()).collect(Collectors.toList());
+        for (Cell ce : horiz) {
```

Comment:
once you have a better data structure - find a fastest way to collect the cells which belong to horizontal line, vertical line and square. and use parallelStreams to find any cell besides the one you are checking if its value has the number - then it returns false